src/interpreter/type_cast.rs

```diff
@@ -12,138 +12,56 @@
 // limitations under the License.
 
 use crate::interpreter::types::{Numeric, PactType, StringLike};
-use core::any::Any;
-use core::convert::TryFrom;
-use std::string::String;
-use std::vec::Vec;
-
-#[derive(Debug, PartialEq)]
-pub enum PactConversionErr {
-    Overflow,
-    UnknownType,
-}
 
-/// A catch-all conversion trait which tries to turn any given `value` into the implementing type
-pub trait AnyTryInto<'a>: Sized {
-    fn any_try_into(value: &'a dyn Any) -> Result<Self, PactConversionErr>;
+/// A blanket trait for conversion into PactType
+pub trait IntoPact<'a, I> {
+    fn into_pact(self) -> PactType<'a>;
 }
 
-/// AnyTryInto implementation for PactType
-impl<'a> AnyTryInto<'a> for PactType<'a> {
-    fn any_try_into(value: &'a dyn Any) -> Result<PactType<'a>, PactConversionErr> {
-        // TODO: refactor the below repetition using macros
-        // Unsigned integer type casting into PactType
-        if let Some(number) = value.downcast_ref::<u8>() {
-            return Ok(PactType::Numeric(Numeric(*number as u64)));
-        }
-        if let Some(number) = value.downcast_ref::<u16>() {
-            return Ok(PactType::Numeric(Numeric(*number as u64)));
-        }
-        if let Some(number) = value.downcast_ref::<u32>() {
-            return Ok(PactType::Numeric(Numeric(*number as u64)));
-        }
-        if let Some(number) = value.downcast_ref::<u64>() {
-            return Ok(PactType::Numeric(Numeric(*number)));
-        }
-        if let Some(number) = value.downcast_ref::<u128>() {
-            return Ok(PactType::Numeric(Numeric(
-                u64::try_from(*number).map_err(|_| PactConversionErr::Overflow)?,
-            )));
-        }
-
-        // String-like type casting into PactType
-        if let Some(string) = value.downcast_ref::<&str>() {
-            return Ok(PactType::StringLike(StringLike(&*string.as_bytes())));
-        }
-        if let Some(string) = value.downcast_ref::<String>() {
-            return Ok(PactType::StringLike(StringLike(string.as_bytes())));
-        }
-        if let Some(string) = value.downcast_ref::<Vec<u8>>() {
-            return Ok(PactType::StringLike(StringLike(&*string)));
-        }
+/// Dummy structs
+struct Numbers;
```

Comment:
💭  Does using these structs affect the usage of `IntoPact` for external crates i.e. will importing the trait outside of this module/crate still give the blanket impls?

Reply:
Yep I'm pointing to this branch from `cennznut-rs` and I can import the trait just fine:
```rust
use pact::interpreter::types::IntoPact;
```

Reply:
Actually the structs seem to be redundant as we can change to:
```rust
impl<'a, T: Into<u64> + Copy> IntoPact<'a, &T> for T { ... }
impl<'a, T: AsRef<[u8]> + ?Sized> IntoPact<'a, &T> for &'a T { ... }
```

I guess I'll remove them :)